pages/chaoss/visualizations/contrib_importance_over_time.py

```diff
@@ -0,0 +1,416 @@
+        time.sleep(1.0)
+        df = cache.grabm(func=ctq, repos=repolist)
+
+    # data ready.
+    start = time.perf_counter()
+    logging.warning(f"{VIZ_ID}- START")
+
+    # test if there is data
+    if df.empty:
+        logging.warning(f"{VIZ_ID} - NO DATA AVAILABLE")
+        return nodata_graph
+    
+    # if the step size is greater than window width raise Alert
+    if step_size > window_width:
+        return dash.no_update, True
+    
+    df_final = process_data(df, 
+                            patterns, 
+                            threshold, 
+                            window_width, 
+                            step_size, 
+                            start_date, 
+                            end_date)
+    
+    # drop rows with NaN values 
+    df_final = df_final.dropna(axis=0)
+    fig = create_figure(df_final, step_size)
+
+    logging.warning(f"{VIZ_ID} - END - {time.perf_counter() - start}")
+    return fig, False
+
+
+def calc_cntrb_importance(df, action_type, threshold):
+    # sort rows in df based on number of contributions from greatest to least
+    df = df.sort_values(by=action_type, ascending=False)
+
+    # calculate the threshold amount of contributions 
+    thresh_cntrbs = df[action_type].sum() * threshold
+    
+    # initilize running sum of contributors who make up contributor importance
+    cntrb_importance = 0
+
+    # initialize running sum of contributions
+    running_sum = 0
+    
+    for _, row in df.iterrows():
+        running_sum += row[action_type] # update the running sum by the number of contributions a contributor has made
+        cntrb_importance+=1 # update contributor importance 
+        # if the running sum of contributions is greater than or equal to the threshold amount, break
+        if running_sum >= thresh_cntrbs:
+            break
+
+    return cntrb_importance
+
+def cntrb_importance_over_time(df, period_from, period_to, window_width, threshold):
+    # subset df such that the rows correspond to the window of time defined by period from and period to
+    mask = (df['created_at'] >= period_from) & (df['created_at'] <= period_to)
+    df_in_range = df.loc[mask]
+    
+    # count the number of contributions each contributor has made according each action type
+    df_count_cntrbs = df_in_range.groupby(['Action', 'cntrb_id'])['cntrb_id'].count().to_frame()
+    df_count_cntrbs = df_count_cntrbs.rename(columns={'cntrb_id': 'count'}).reset_index()
+     
+    # pivot df such that the column names correspond to the different action types and the values are the number of contributions of each contributor
+    df_count_cntrbs = df_count_cntrbs.pivot(columns='Action', values='count')
+
+    # initilize empty df with columns that correspond to the different action types to store contributor importance 
+    df_cntrb_importance = pd.DataFrame(columns=df_count_cntrbs.columns)
+    
+    # create a dictionary to store contributor importance for each action type
```

Comment:
If you want to follow line for line same logic as staling_issues, store as list instead of dictionary, have for loop be on the list of actions in a set order